libsignify/src/key.rs

```diff
@@ -162,14 +162,35 @@ impl PrivateKey {
                     return Err(Error::BadPassword);
                 }
 
-                // Confirmed the decryption worked, mutating the key structure.
-                self.complete_key = decrypted_key.0;
+                match PrivateKey::verify_keypair(&decrypted_key.0) {
+                    Ok(()) => {
+                        // Confirmed the decryption worked and the keys are matching.
+                        self.complete_key = decrypted_key.0;
+                    }
+                    Err(e) => return Err(e),
+                }
+
                 Ok(())
             }
             Err(e) => Err(e),
         }
     }
 
+    pub(crate) fn into_keypair(complete_key: &[u8; FULL_KEY_LEN]) -> ed25519_dalek::Keypair {
+        // This `unwrap` is erased in release mode.
+        ed25519_dalek::Keypair::from_bytes(complete_key).unwrap()
```

Comment:
I don't think this unwrap is free? This gets passed user-controlled data, so can actually panic, as the error shows, doesn't it?

Reply:
It should be iiuc. `Keypair::from_bytes` [only errors](https://docs.rs/ed25519-dalek/latest/src/ed25519_dalek/keypair.rs.html#80-91) when the length of the slice is wrong. It doesn't actually parse the bytes as curve points. Since `FULL_KEY_LEN` is a constant, the optimizer can remove the error branch entirely since it knows it would never be taken.

Reply:
This is so long ago I don't know why I thought it might be causing problems 🤔

Reply:
I ended up being wrong here (what a surprise /s ) since I tunnel visioned on this. The public key is validated to the extent that it "must be a compressed curve point" but not much more. This unwrap is gone now since, like you said, its in the path of fresh user-controlled data that hasn't been looked at yet.